pytket/pytket/rmbqc/mpattern.py

```diff
@@ -0,0 +1,533 @@
+        
+        :param c:       An arbitrary pytket circuit that we want converted to 
+                            a measurement pattern.
+        :param type:    Circuit
+        """
+        self.c = c
+        
+    def label_squish(self, g: GraphS, io_map: Dict[str, Dict[Qubit, int]]) -> None:
+        """
+        Updates the input/output labels of the MPattern to matched a squished
+        graph caused by g.copy(). The reason why we use g.copy() is because the
+        easiest way to split a zx diagram into multiple disjoint zx.diagrams is
+        to first identify X disjoint clusters of qubits, the copy the original
+        graph into X graphs, and then for each of the copies remove all the
+        vertices outside the corresponding cluster.
+        
+        :param g:       A pyzx graph representing a zx diagram.
+        :param type:    GraphS
+        
+        :param io_map:  A dictionary containing the current input/output mappings
+                            which map the input/output qubits of the original circuit
+                            to the corresponding vertices in the ZX diagram.
+        :param type:    Dict[str,Dict[Qubit,int]]
+        """
+        original_labels = sorted(
+            list(g.vertices())
+        )  # Sort the original vertices in ascending order.
+        # The following code will update the i/o map with the new labels of the
+        # input/output qubits after the squish.
+        for i in io_map["i"].keys():
+            for v in range(len(original_labels)):
+                if io_map["i"][i] == original_labels[v]:
+                    io_map["i"][i] = v
+                    break
+        for o in io_map["o"].keys():
+            for v in range(len(original_labels)):
+                if io_map["o"][o] == original_labels[v]:
+                    io_map["o"][o] = v
+                    break
+                
+    def remove_redundant(self, g: GraphS, io_map: Dict[str, Dict[Qubit, int]]) -> None:
+        """
+        Removes simples edges from a zx diagram by merging the connected
+        vertices.
+        
+        :param g:       A zx diagram with some remaining simple edges we want to remove.
+        :param type:    GraphS
+        
+        :param io_map:  A dictionary containing the current input/output mapping
+                            to map the input/output qubits of the original circuit
+                            to their corresponding vertices in the ZX diagram.
+        :param type:    Dict[str,Dict[Qubit,int]]
+        """
+        simple_edges = set()
+        for edge in g.edge_set():
+            if g.edge_type(edge) == 1:
+                simple_edges.add(edge)
+        for edge in simple_edges:
+            v1, v2 = edge
+            g.remove_edge(edge)
+            is_boundary = True
+            removing_input = True
+            remove_vertex = v1
+            keep_vertex = v2
+            if v1 in g.inputs:
```

Comment:
Tbh I think they are not actually necessary... Basically when we merge the vertices we just move all the edges on one of the vertices to the other, add the phase of the one we are removing to the one we are keeping, and then just delete it.

The only other thing we need to do is make sure that if the vertex we removed was designated as an input or output, then the other vertex must take its place in the corresponding maps.

Beyond that, it shouldn't matter which vertex we remove and which one we keep. I'm not sure why I wrote the function the way I have, it might have had to do with just visualisation reasons - because I was drawing the zx diagrams I think I may have just wanted to prioritize keeping the vertices that would lead to clearer looking diagrams so I could better see what was going on.

So this is definitely the sort of thing that could be re-written and improved... However, if I were to try to change it right now I might make a mess.

The thing is, the first thing Ill do when I resume the PhD is probably the tri-planar zx stuff and that means that this is one of the first functions to be re-written so I suggest we leave it till then - but if you want it done now I can try hacking at it.